### PR TITLE
Let reference input be either a string or a FIle

### PR DESCRIPTION
**Pull request by tmooney**
This is an updated version of the idea first presented in #682 (and potentially addresses #674).

I wanted to get this PR open, but we shouldn't merge this quite yet.  Using this version of the workflow requires a patch to Cromwell.  I have a change that makes it work as before when a `string` is supplied, but I'm still testing further changes to make Cromwell work correctly when a `File` is used instead.

### CODE REVIEW

File: definitions/tools/varscan_somatic.cwl [hunk 1]
```diff
@@ -21,7 +21,10 @@ inputs:
             position: 2
         secondaryFiles: [^.bai]
     reference:
-        type: string
+        type:
+            - string
+            - File
+        secondaryFiles: [.fai, ^.dict]
```
Comment (chrisamiller) [inline definitions/tools/varscan_somatic.cwl]: Do all of these actually need the `.dict` file?  I don't relish the thought of digging through all 80-something of these to verify that, but I'm pretty sure, for example, that varscan here does not.

File: definitions/tools/filter_vcf_mapq0.cwl [hunk 2]
```diff
@@ -23,15 +23,18 @@ inputs:
             position: 2
         secondaryFiles: [.bai]
     reference: 
-         type: string
-         inputBinding:
-             position: 3
+        type:
+            - string
+            - File
+        secondaryFiles: [.fai, ^.dict]
```
Comment (chrisamiller) [inline definitions/tools/filter_vcf_mapq0.cwl]: This is another one that definitely doesn't need the .dict file.

**Comment by serge2016** [comment]
Thank you for implementing this!

**Comment by tmooney** [comment]
@chrisamiller agreed, this was definitely a conservative approach to identifying which secondaryFiles to pass along.  Curiously, some workflow engines pass along all the secondary files from the original input, anyway, even if each tool doesn't explicitly request them, so it may be harder than expected to test which ones are needed.

I've just rebased this PR to account for it sitting a few months.  Now that I've got a new docker environment building it comes time where it'd be nice to sort this out.  Would you mind giving a re-review?  (Any other pipelines you'd like to see receive this treatment? For this PR I only focused on the main DNA pipelines that are all intertwined tool-wise.  It's generally safe for an unconverted pipeline to use a converted tool, though.)

**Comment by tmooney** [comment]
I'd be happy to receive feedback on which `secondaryFiles` are needed. If there are any you're aware of off-hand, I could add it to this PR (along with the one you already mentioned).  However, I'm leaning towards getting this pushed and sorting it out more comprehensively in a follow-up.

(As long as we're using the `string` version of the input locally, what's listed is of zero consequence to our pipelines.  It's when we're running "in the cloud" that this would come into play.)

**Comment by serge2016** [comment]
Dear colleagues, could you show an example of `*.index` file? You have added it as a secondary with reference fasta, but it doesn't exist in example data.

**Comment by tmooney** [comment]
The `.index` file is generated by VEP:
https://useast.ensembl.org/info/docs/tools/vep/script/vep_cache.html#fasta

If you need to generate this `.index` file, this issue might be useful: Ensembl/ensembl-vep#211.

**Comment by serge2016** [comment]
@tmooney, thank you! But those variants produce different `.index` files. Witch one is correct?
What should be in this `.index` file?

**Comment by tmooney** [comment]
Sounds like it is better to use the htslib version rather than the older `Bio::DB::Fasta` version.  If nothing else I'd use an `.index` file produced by VEP itself after running it with a FASTA file; that's the most sure way to get one that will work.

It's functionally very similar to the `.fai` file--a quick way to find out where in the file each contig begins.  It needs to be in a format that VEP understands, which may vary depending on which modules are available to a particular install of VEP.  Thus it may be good to generate the file from within a Docker container running on the image used in the workflow, currently `mgibio/vep_helper-cwl:1.1.0`.

**Comment by jasonwalker80** [comment]
I would add that our workflow is based on the implementation we have setup, ie. the `mgibio/vep_helper-cwl:1.1.0` docker image. The `.index` just happens to be the file that is generated when using that image. I agree the solution to the missing input is to run VEP once on the FASTA and let it generate the secondary file that is needed. As for the test data, we should include this step as an instruction to generated the secondary file before attempting to run the tests.

**Comment by serge2016** [comment]
@tmooney, VEP doesn't produce `.index` file... That's a problem, I can't solve. VEP only produces `.fai`.

**Comment by tmooney** [comment]
My apologies; I mixed up the two index formats.  The `.index` file comes from the older [`Bio::DB::Fasta`](https://metacpan.org/pod/release/CJFIELDS/BioPerl-1.5.9_1/Bio/DB/Fasta.pm) module rather than the newer version.  As you note, when htslib is available VEP prefers to use the newer index format instead and does not require or produce this `.index` file.

Perhaps the right course of action here is for us to remove the `.index` secondaryFile requirement from our pipelines.  As long as a new enough VEP with the proper modules is available, one could substitute an empty `.index` file for now.

**Comment by serge2016** [comment]
@tmooney, thank you! For now, we use http://web.mit.edu/meme_v4.11.4/share/doc/fasta-make-index.html to generate `.index` file.

It would be great, if `.index` is removed from required files! Is it possible to make it optional secondary file?

**Comment by tmooney** [comment]
@serge2016 PR #835 is open to remove this.  In CWL v1.1 there is a way to specify optional secondaryFiles (adding a `?` to the end of the extension), but we're not ready to migrate our workflows to this version yet.

**Pull request merged**